mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp

```diff
@@ -1124,15 +1124,15 @@ static void getGenericEffectsImpl(
         &effects,
     LinalgOp linalgOp) {
   for (auto [index, operand] : llvm::enumerate(linalgOp.getDpsInputs())) {
-    if (!llvm::isa<MemRefType>(operand.getType()))
+    if (!operand || !llvm::isa<MemRefType>(operand.getType()))
```

Comment:
Why are we checking if operand is a `nullptr`? Shouldn't the `inputOperand` and `outputOperand` be non-null and whoever is removing this operand should removet from these lists as well?

Reply:
Since we remove dead values in chronological order, we may have removed `inputOperand` or `outputOperand` when checking the memory effect of `linalgOp`s.
We should have done liveness analysis before removing those operands, so the removed operands are indeed dead values.
Does this sound reasonable?

Reply:
Yeah, if an operand can suddenly point to null values, then that is a problem and a lot of code can break. I am not sure how you are removing values without knowing they are indeed dead.

Reply:
> Since we remove dead values in chronological order, we may have removed inputOperand or outputOperand when checking the memory effect of linalgOps.

I would think that should assert: the destructor of a Value is checking for use_empty() I believe.

Reply:
```mlir
%cst_2 = arith.constant dense<[[1., 2., 3.], [4., 5., 6.]]> : tensor<2x3xf64>
%cst_3 = arith.constant dense<0.> : tensor<3x2xf64>
%transposed = linalg.transpose ins(%cst_2 : tensor<2x3xf64>) outs(%cst_3 : tensor<3x2xf64>) permutation = [1, 0]
return
```
In example above, `%cst_2`, `%cst_3`, `transposed` are dead values since they are not used, however, we drop uses and erase  these values in order and turning the content of the region into something similar to snippet below.

```
%transposed = linalg.transpose ins(nullptr) outs(nullptr) permutation = [1, 0]
return
```

We've checked the liveness of values before removing them, so I think we should not assert while removing dead values.

If we don't want to check whether arguments have been removed, we may need to change the removing order into reverse chronological order.